format/rtsp/client.go

```diff
@@ -0,0 +1,1173 @@
+			self.session = fields[0]
+		}
+	}
+	if res.StatusCode == 401 {
+		if err = self.handle401(res); err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (self *Client) handle401(res *Response) (err error) {
+	/*
+		RTSP/1.0 401 Unauthorized
+		CSeq: 2
+		Date: Wed, May 04 2016 10:10:51 GMT
+		WWW-Authenticate: Digest realm="LIVE555 Streaming Media", nonce="c633aaf8b83127633cbe98fac1d20d87"
+	*/
+	authval := res.Headers.Get("WWW-Authenticate")
+	hdrval := strings.SplitN(authval, " ", 2)
+	var realm, nonce string
+
+	if len(hdrval) == 2 {
+		for _, field := range strings.Split(hdrval[1], ",") {
+			field = strings.Trim(field, ", ")
+			if keyval := strings.Split(field, "="); len(keyval) == 2 {
+				key := keyval[0]
+				val := strings.Trim(keyval[1], `"`)
+				switch key {
+				case "realm":
+					realm = val
+				case "nonce":
+					nonce = val
+				}
+			}
+		}
+
+		if realm != "" {
+			var username string
+			var password string
+
+			if self.url.User == nil {
+				err = fmt.Errorf("rtsp: no username")
+				return
+			}
+			username = self.url.User.Username()
+			password, _ = self.url.User.Password()
+
+			self.authHeaders = func(method string) []string {
+				var headers []string
+				if nonce == "" {
+					headers = []string{
+						fmt.Sprintf(`Authorization: Basic %s`, base64.StdEncoding.EncodeToString([]byte(username+":"+password))),
+					}
+				} else {
+					hs1 := md5hash(username + ":" + realm + ":" + password)
+					hs2 := md5hash(method + ":" + self.requestUri)
+					response := md5hash(hs1 + ":" + nonce + ":" + hs2)
+					headers = []string{fmt.Sprintf(
+						`Authorization: Digest username="%s", realm="%s", nonce="%s", uri="%s", response="%s"`,
+						username, realm, nonce, self.requestUri, response)}
+				}
+				return headers
+			}
+		}
+	}
+
+	return
+}
+
+func (self *Client) findRTSP() (block []byte, data []byte, err error) {
+	const (
+		R = iota + 1
+		T
+		S
+		Header
+		Dollar
+	)
+	var _peek [8]byte
+	peek := _peek[0:0]
+	stat := 0
+
+	for i := 0; ; i++ {
+		var b byte
+		if b, err = self.brconn.ReadByte(); err != nil {
+			return
+		}
+		switch b {
+		case 'R':
+			if stat == 0 {
+				stat = R
+			}
+		case 'T':
+			if stat == R {
+				stat = T
+			}
+		case 'S':
+			if stat == T {
+				stat = S
+			}
+		case 'P':
+			if stat == S {
+				stat = Header
+			}
+		case '$':
+			if stat != Dollar {
+				stat = Dollar
+				peek = _peek[0:0]
+			}
+		default:
+			if stat != Dollar {
+				stat = 0
+				peek = _peek[0:0]
+			}
+		}
+
+		if false && self.DebugRtp {
+			fmt.Println("rtsp: findRTSP", i, b)
+		}
+
+		if stat != 0 {
+			peek = append(peek, b)
+		}
+		if stat == Header {
+			data = peek
+			return
+		}
+
+		if stat == Dollar && len(peek) >= 12 {
+			if self.DebugRtp {
+				fmt.Println("rtsp: dollar at", i, len(peek))
+			}
+			if blocklen, _, ok := self.parseBlockHeader(peek); ok {
+				left := blocklen + 4 - len(peek)
+				block = append(peek, make([]byte, left)...)
```

Comment:
this potential out of range make slice
you can see this issue
https://github.com/golang/go/issues/38673

Reply:
sure, but is that the only change required to merge this PR?

Reply:
I think conflict file must be fixed too

Reply:
sure, but is that really the only thing? I want to make sure that it is worth my time to resolve this and you can merge it.

Reply:
for merge just @nareix have role, I not have the role
but we can work together testing this branch

Reply:
sure, but I would rather address all comments at once so could @nareix also have a look and add any other comments so I can fix all issues if any.

I have been running this for a while now and haven't gotten any issues

Reply:
in my case I using [aler9/rtsp-simple-server](https://github.com/aler9/rtsp-simple-server) 
and publish with rtmp protocol to rtsp-simple-server with bitrate 4000 Kbps and I connect with rtsp client with this library,to rtsp-simple-server and after 30-1 hours I got an error `makeslice: len out of range`

Reply:
Had a quick look and I think this part of the code is unchanged from joy4 and it will be difficult to troubleshoot. Can you maybe try troubleshooting on your end to find out why this happens and suggest a fix?